levels.go

```diff
@@ -1001,7 +1001,15 @@ func (s *levelsController) addSplits(cd *compactDef) {
 			return
 		}
 		if i%N == N-1 {
-			addRange(t.Biggest())
+			// Right should always have ts=maxUint64 otherwise we'll lose keys
+			// in subcompaction. Consider the following.
+			// Top table is [A1...C3(deleted)]
+			// bot table is [B1....C2]
+			// This will generate splits like [A1 ... C2] . Notice that we
+			// dropped the C3 which is the last key of the top table.
+			// See TestCompaction/with_split test.
+			right := y.KeyWithTs(y.ParseKey(t.Biggest()), math.MaxUint64)
```

Comment:
getKeyRange takes care of these things. I thought I was using it -- that's the reason for the bug.

In fact, don't we want to consider the right most key? In that case, the right should be using 0 as the version.

Reply:
We're generating key ranges using the `bot` tables so they don't have the correct key version.
>In fact, don't we want to consider the right most key? In that case, the right should be using 0 as the version.

No. We skip the right keys of the key range. 
https://github.com/dgraph-io/badger/blob/6c074551f8f5a5b470357236cd90e860d0deb517/levels.go#L666-L668